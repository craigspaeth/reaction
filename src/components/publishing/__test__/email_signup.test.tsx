import { mount } from "enzyme"
import "jest-styled-components"
import * as React from "react"
import * as renderer from "react-test-renderer"
import request from "request"
import EmailSignup from "../email_signup"

jest.mock("request", () => {
  return { post: jest.fn() }
})
jest.useFakeTimers()

describe("EmailSignup", () => {
  it("renders an email signup", () => {
    const emailSignup = renderer.create(<EmailSignup signupUrl="#" />)
    expect(emailSignup).toMatchSnapshot()
  })

  it("submits an email and removes itself when successful", () => {
    const viewer = mount(<EmailSignup signupUrl="#" />)
    viewer.setState({ value: "foo@goo.net" })
    viewer.find("button").simulate("click")
    expect(request.post).toBeCalled()
    expect(request.post.mock.calls[0][0].body.email).toEqual("foo@goo.net")
    request.post.mock.calls[0][1]()
    const state = viewer.state()
    expect(state.message).toEqual("Thank you!")

    jest.runAllTimers()
    const postTimeoutState = viewer.state()
    expect(postTimeoutState.submitted).toBe(true)
    expect(postTimeoutState.disabled).toBe(false)
    expect(postTimeoutState.message).toEqual("")
  })

  it("handles signup errors", () => {
    const viewer = mount(<EmailSignup signupUrl="#" />)
    viewer.setState({ value: "foo@goo.net" })
    viewer.find("button").simulate("click")
    expect(request.post).toBeCalled()
    request.post.mock.calls[1][1]("Error")
    const state = viewer.state()
    expect(state.message).toEqual("Error. Please try again")
    expect(state.error).toBe(true)

    jest.runAllTimers()
    const postTimeoutState = viewer.state()
    expect(postTimeoutState.disabled).toBe(false)
    expect(postTimeoutState.error).toBe(false)
    expect(postTimeoutState.message).toEqual("")
  })

  it("validates email addresses", () => {
    const viewer = mount(<EmailSignup signupUrl="#" />)
    viewer.setState({ value: "foo" })
    viewer.find("button").simulate("click")
    const state = viewer.state()
    expect(state.message).toEqual("Invalid Email... Please try again")
    expect(state.error).toBe(true)

    jest.runAllTimers()
    const postTimeoutState = viewer.state()
    expect(postTimeoutState.disabled).toBe(false)
    expect(postTimeoutState.error).toBe(false)
    expect(postTimeoutState.message).toEqual("")
  })
})
