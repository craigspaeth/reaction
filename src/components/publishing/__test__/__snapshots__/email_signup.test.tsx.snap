// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailSignup renders an email signup 1`] = `
.c5 {
  background: #e5e5e5;
  color: black;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 15px 30px;
  font-size: 13px;
  line-height: 1;
  outline: 0;
  -webkit-transition: background-color .25s,color .25s;
  transition: background-color .25s,color .25s;
  margin: 10px;
  border: none;
  box-sizing: border-box;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: 'ITC Avant Garde Gothic W04',  'AvantGardeGothicITCW01D 731075',  'AvantGardeGothicITCW01Dm',  'Helvetica',  'sans-serif';
  -webkit-font-smoothing: antialiased;
  -webkit-text-transform: uppercase;
  text-transform: uppercase;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
}

.c5:hover:not(:disabled) {
  cursor: pointer;
  background: #cccccc;
}

.c3 {
  padding: 10px;
  box-shadow: none;
  font-size: 17px;
  -webkit-transition: border-color .25s;
  transition: border-color .25s;
  margin-right: 10px;
  resize: none;
  font-family: 'Adobe Garamond W08',  'adobe-garamond-pro',  'AGaramondPro-Regular',  'Times New Roman',  'Times',  'serif';
  -webkit-font-smoothing: antialiased;
  border: 2px solid #e5e5e5;
  -webkit-transition: border-color .25s;
  transition: border-color .25s;
  width: 100%;
  border-width: 1px;
  color: black;
}

.c3:focus,
.c3.focused {
  border-color: #6e1fff;
  outline: 0;
}

.c3:disabled {
  border: 2px dotted #e5e5e5;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  max-width: 350px;
  width: 100%;
  margin-bottom: 40px;
}

.c1 {
  font-family: Unica77LLWebMedium ,      'Arial',      'serif';
  -webkit-font-smoothing: antialiased;
  font-size: 19px;
  line-height: 1.5em;
  margin-bottom: 10px;
}

.c4 {
  background: black;
  color: white;
  border-radius: 2px;
  height: 30px;
  width: 80px;
  margin-left: -100px;
  font-family: 'ITC Avant Garde Gothic W04',  'AvantGardeGothicITCW01D 731075',  'AvantGardeGothicITCW01Dm',  'Helvetica',  'sans-serif';
  -webkit-font-smoothing: antialiased;
  -webkit-text-transform: uppercase;
  text-transform: uppercase;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  font-size: 11px;
  line-height: 1.65em;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
}

.c4:hover:not(:disabled) {
  background: #6e1fff;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    Stay up to date with Artsy Editorial
  </div>
  <div
    className="c2"
  >
    <input
      className="c3"
      onChange={[Function]}
      placeholder="Enter Your Email..."
      readOnly={false}
      type="email"
      value=""
    />
    <button
      className="c4 c5"
      disabled={false}
      onClick={[Function]}
    >
      <span>
        Subscribe
      </span>
    </button>
  </div>
</div>
`;
